File: examples/go/hand/agg/handlers/post_blind.go
package handlers

import (
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardPostBlind(state HandState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Hand does not exist")
	}
	if state.IsComplete() {
		return angzarr.NewCommandRejectedError("Hand already complete")
	}
	return nil
}

func validatePostBlind(cmd *examples.PostBlind, state HandState) (*PlayerHandState, error) {
	player := state.GetPlayerByRoot(cmd.PlayerRoot)
	if player == nil {
		return nil, angzarr.NewCommandRejectedError("Player not in hand")
	}
	if cmd.Amount <= 0 {
		return nil, angzarr.NewCommandRejectedError("Amount must be positive")
	}
	return player, nil
}

func computeBlindPosted(cmd *examples.PostBlind, state HandState, player *PlayerHandState) *examples.BlindPosted {
	actualAmount := cmd.Amount
	if actualAmount > player.Stack {
		actualAmount = player.Stack
	}

	newStack := player.Stack - actualAmount
	newPot := state.TotalPot() + actualAmount

	return &examples.BlindPosted{
		PlayerRoot:  cmd.PlayerRoot,
		BlindType:   cmd.BlindType,
		Amount:      actualAmount,
		PlayerStack: newStack,
		PotTotal:    newPot,
		PostedAt:    timestamppb.New(time.Now()),
	}
}

// HandlePostBlind handles the PostBlind command.
func HandlePostBlind(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state HandState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.PostBlind
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardPostBlind(state); err != nil {
		return nil, err
	}
	player, err := validatePostBlind(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computeBlindPosted(&cmd, state, player)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}

File: examples/rust/hand/agg/src/handlers/post_blind.rs
//! PostBlind command handler.

use angzarr_client::proto::examples::{BlindPosted, PostBlind};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::{HandState, PlayerHandState};

fn guard(state: &HandState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }
    Ok(())
}

fn validate<'a>(cmd: &PostBlind, state: &'a HandState) -> CommandResult<&'a PlayerHandState> {
    let player = state
        .get_player(&cmd.player_root)
        .ok_or_else(|| CommandRejectedError::new("Player not in hand"))?;

    if cmd.amount <= 0 {
        return Err(CommandRejectedError::new("Amount must be positive"));
    }

    Ok(player)
}

fn compute(cmd: &PostBlind, state: &HandState, player: &PlayerHandState) -> BlindPosted {
    let actual_amount = cmd.amount.min(player.stack);
    let new_stack = player.stack - actual_amount;
    let new_pot = state.total_pot() + actual_amount;

    BlindPosted {
        player_root: cmd.player_root.clone(),
        blind_type: cmd.blind_type.clone(),
        amount: actual_amount,
        player_stack: new_stack,
        pot_total: new_pot,
        posted_at: Some(angzarr_client::now()),
    }
}

pub fn handle_post_blind(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: PostBlind = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    let player = validate(&cmd, state)?;

    let event = compute(&cmd, state, player);
    let event_any = pack_event(&event, "examples.BlindPosted");

    Ok(new_event_book(command_book, seq, event_any))
}
